pkg/provisioner/ironic/ironic.go

```diff
@@ -1760,34 +1761,62 @@ func (p *ironicProvisioner) RemoveBMCEventSubscriptionForNode(subscription metal
 	return operationComplete()
 }
 
-// TODO(hroyrh) : Replace with GetDataImageStatus function once the virtua_media.get
-// api is available.
-// Checks if the last VirtualMedia action(attach/detach) to a BareMetalHost was
-// successful of not.
-func (p *ironicProvisioner) IsDataImageReady() (isNodeBusy bool, nodeError error) {
-	// TODO(hroyrh)
-	// Get BareMetalHost VirtualMedia details using a GET api to vmedia
-
-	// Check if Ironic API version supports DataImage API
-	// Needs version >= 1.89
-	if !p.availableFeatures.HasDataImage() {
-		return true, fmt.Errorf("ironic version=%d doesn't support DataImage API, needs version>=1.89", p.availableFeatures.MaxVersion)
+// Uses the Ironic Virtual Media Get API which synchronously fetches the
+// virtual media details for the given node.
+// We return only the bool isImageAttached because the url in the response
+// body is always the same - based on the node uuid, so it is not useful
+// for any comparison purpose.
+func (p *ironicProvisioner) GetDataImageStatus() (isImageAttached bool, err error) {
+	// Check if Ironic API version supports Virtual Media Get API
+	// Needs version >= 1.93
+	if !p.availableFeatures.HasVirtualMediaGetAPI() {
+		return false, fmt.Errorf("ironic version=%d doesn't support Virtual Media Get API, needs version>=1.93", p.availableFeatures.MaxVersion)
 	}
 
 	node, err := p.getNode()
 	if err != nil {
 		return true, err
 	}
 
-	isNodeBusy = node.Reservation != ""
+	// If the node is busy, return error, since the get request will anyways
+	// fail
+	// TODO(hroyrh): is there a better way to do this ?
+	isNodeBusy := node.Reservation != ""
+	if isNodeBusy {
+		p.log.Info("node is busy, retry GetVirtualMedia request in some time", "nodeid", node.UUID)
+		return false, fmt.Errorf("node is busy")
+	}
+
+	result := nodes.GetVirtualMedia(p.ctx, p.client, p.nodeID)
+	err = result.Err
+	if err != nil {
+		return false, err
+	}
+
+	body := result.Body
 
-	// In case the error node encountered was related to something else
-	// TODO(hroyrh) : Is this check valid ?
-	if !strings.Contains(node.LastError, "attach") && !strings.Contains(node.LastError, "detach") {
-		return isNodeBusy, nil
+	data, err := json.Marshal(body)
+	if err != nil {
+		p.log.Info("GetDataImage, failed to marshal response body", "Error", err)
+		return false, err
+	}
+
+	var jsonData []map[string]interface{}
```

Comment:
Let's create a list of structures, not maps. We know how the response is supposed to look.

Reply:
I've created a struct for VirtualMedia, will move it to gophercloud next.